workspace/TS100/Core/Src/power.cpp

```diff
@@ -24,6 +25,21 @@ int32_t tempToX10Watts(int32_t rawTemp) {
 
 void setTipX10Watts(int32_t mw) {
 	int32_t output = X10WattsToPWM(mw, 1);
+	if (fastPWM && output == powerPWM) {
```

Comment:
Is the purpose of this logic to run the PWM faster when we are doing the inital heat up to allow for finer controll and then slow down after this point in time?

(Would appreciate some form of comment block ahead of this as to why this approach was chosen :)

Reply:
The idea is that the faster we run PWM, the less averaging we require from the output of the PSUs. But we need 7 ms before each ADC measurement cycle, and few more ms to do the 4 measurements, and we need to do that after each PWM cycle. This whole ADC time is "wasted" because the tip doesn't receive power (and we want the tip to receive as much as possible in the power-unlimited case), so naturally we'd like to be able to do the ADC measurements less often (by having longer PWM cycles). So my idea is to switch to fast PWM when the conditions do not require anything close to full power.

Reply:
Another idea is that we could totally "skip" every other ADC cycle during large heatup events to save the dead time, and when heating up large differences we are not doing a _huge_ amount of logic in the PID (just saturating at the output).

Of course this may have implications in stability, but _could_ be an interesting optimisation.

But yes I agree with this concept :)

Reply:
I had that another idea too but couldn't come up with a simple enough implementation as it would involve changing the PWM period every single cycle and probably other trickery and would affect power/pwm calculations etc, so I got too confused to try to implement it :)

Reply:
Absolutely agree it could be very confusing 😅 
I'm happy to go with this implementation for now.